File: native/Android/ui/ProduceViewModel.kt
package com.example.produceapp.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.produceapp.data.*
import com.example.produceapp.network.ProduceService
import com.example.produceapp.util.Resource
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

// =============================================================================
// ProduceViewModel.kt — Android MVVM 架構中的 ViewModel 層
//
// 架構說明：
//   MVVM 資料流：
//   UI (Composable) → 呼叫 ViewModel 方法 → ViewModel 呼叫 ProduceService
//   → API 回應 → ViewModel 更新 StateFlow → UI 自動重組 (Recomposition)
//
//   StateFlow vs LiveData：
//   - StateFlow 是 Kotlin Coroutines 的響應式資料容器，生命週期感知更好
//   - collectAsState() 讓 Composable 訂閱 StateFlow，資料變更時自動重新渲染
//   - MutableStateFlow (私有) + StateFlow (公開) = 封裝原則，
//     只有 ViewModel 自身可以修改狀態，UI 只能讀取
//
// Hilt 依賴注入：
//   @HiltViewModel + @Inject constructor 讓 Hilt 自動管理 ViewModel 的建立與依賴注入
//   ProduceService 由 Hilt 在 AppModule 中提供，不需要手動 new
//
// 對應 iOS ProduceViewModel.swift：
//   @MainActor class ProduceViewModel     = @HiltViewModel class ProduceViewModel
//   @Published var dailyPrices            = MutableStateFlow _dailyPrices
//   enum Resource<T>                      = sealed class Resource<T>
//   Task { await fetchDashboardData() }   = init { fetchDashboardData() }
// =============================================================================

/**
 * 農產品資料 ViewModel
 * 使用 Hilt 依賴注入，由 Hilt 管理生命週期（與 Activity/Fragment 綁定）
 */
@HiltViewModel
class ProduceViewModel @Inject constructor(
    // ProduceService 由 Hilt AppModule 提供
    // 注意：目前直接注入 ProduceService，建議未來改為注入 ProduceRepository 以支援離線快取
    private val produceService: ProduceService
) : ViewModel() {

    // MARK: - StateFlow 狀態定義
    // 每個 MutableStateFlow 對應一個 API 請求的三態狀態機
    // 初始值均為 Resource.Loading()，讓 UI 立即顯示骨架屏

    /** 我的收藏清單 + 是否達到目標提醒價格 */
    private val _favorites = MutableStateFlow<Resource<List<FavoriteAlertDto>>>(Resource.Loading())
    val favorites: StateFlow<Resource<List<FavoriteAlertDto>>> = _favorites

    /** 今日農產品批發價格列表（分頁，預設第 1 頁） */
    private val _dailyPrices = MutableStateFlow<Resource<List<ProduceDto>>>(Resource.Loading())
    val dailyPrices: StateFlow<Resource<List<ProduceDto>>> = _dailyPrices

    /** 今日交易量前 10 名農產品 */
    private val _topVolume = MutableStateFlow<Resource<List<ProduceDto>>>(Resource.Loading())
    val topVolume: StateFlow<Resource<List<ProduceDto>>> = _topVolume

    /** 單日漲幅超過 50% 的價格異常警告清單 */
    private val _anomalies = MutableStateFlow<Resource<List<PriceAnomalyDto>>>(Resource.Loading())
    val anomalies: StateFlow<Resource<List<PriceAnomalyDto>>> = _anomalies

    /** 高麗菜（LA1）歷史 30 天價格，供折線圖展示 */
    private val _historicalData = MutableStateFlow<Resource<List<HistoricalPriceDto>>>(Resource.Loading())
    val historicalData: StateFlow<Resource<List<HistoricalPriceDto>>> = _historicalData

    /** 明日預測價格（由 7 日移動平均線趨勢計算） */
    private val _predictedData = MutableStateFlow<Resource<List<PricePredictionDto>>>(Resource.Loading())
    val predictedData: StateFlow<Resource<List<PricePredictionDto>>> = _predictedData

    /** 當季盛產農作物清單 */
    private val _seasonalCrops = MutableStateFlow<Resource<List<SeasonalCropDto>>>(Resource.Loading())
    val seasonalCrops: StateFlow<Resource<List<SeasonalCropDto>>> = _seasonalCrops

    /** 颱風/豪大雨天氣預警 */
    private val _weatherAlerts = MutableStateFlow<Resource<WeatherAlertDto>>(Resource.Loading())
    val weatherAlerts: StateFlow<Resource<WeatherAlertDto>> = _weatherAlerts

    /** 今日省錢食譜推薦 */
    private val _budgetRecipes = MutableStateFlow<Resource<List<BudgetRecipeDto>>>(Resource.Loading())
    val budgetRecipes: StateFlow<Resource<List<BudgetRecipeDto>>> = _budgetRecipes

    // MARK: - 初始化

    /**
     * ViewModel 建立時自動載入首頁所需的所有資料
     * viewModelScope 確保 Coroutine 在 ViewModel 銷毀時自動取消，避免記憶體洩漏
     */
    init {
        fetchDashboardData()
    }

    // MARK: - 資料載入方法

    /**
     * 載入首頁所需的所有資料區塊
     * 使用單一 try-catch 包裝所有請求，任一失敗則所有狀態都設為 Error
     *
     * 注意：目前是循序執行（一個完成才執行下一個）。
     * 未來優化方向：使用 async { } + awaitAll() 讓多個 API 並行執行，縮短載入時間。
     */
    fun fetchDashboardData() {
        viewModelScope.launch {
            try {
                // 1. 載入價格異常警告（漲幅 > 50% 的農產品）
                _anomalies.value = Resource.Success(produceService.getPriceAnomalies())

                // 2. 載入今日熱門交易（交易量前 10 名）
                _topVolume.value = Resource.Success(produceService.getTopVolumeCrops())

                // 3. 載入當季盛產日曆
                _seasonalCrops.value = Resource.Success(produceService.getSeasonalCrops())

                // 4. 載入今日菜價（第 1 頁，20 筆）
                val response = produceService.getDailyPrices("", 1, 20)
                _dailyPrices.value = Resource.Success(response.data)

                // 5. 載入高麗菜（LA1）歷史趨勢圖表資料
                //    高麗菜是台灣消費量最大的葉菜類，作為首頁示範圖表
                val history = produceService.getPriceHistory("LA1")
                _historicalData.value = Resource.Success(history)

                // 6. 計算明日預測價格
                //    Up   → 預測比近 7 日均價高 5%
                //    Down/Stable → 預測比近 7 日均價低 5%
                val forecast = produceService.getPriceForecast("LA1")
                val nextDayPrice = if (forecast.trend == "Up") {
                    forecast.recentAverage * 1.05
                } else {
                    forecast.recentAverage * 0.95
                }
                _predictedData.value = Resource.Success(listOf(
                    PricePredictionDto("Next Day", nextDayPrice)
                ))

                // 7. 載入我的收藏及是否達到目標提醒價
                _favorites.value = Resource.Success(produceService.getFavorites())

            } catch (e: Exception) {
                // 任一 API 失敗時，將所有狀態設為 Error
                e.printStackTrace()
                val errorMsg = e.message ?: "網路連線失敗，請稍後再試"
                _anomalies.value      = Resource.Error(errorMsg)
                _topVolume.value      = Resource.Error(errorMsg)
                _seasonalCrops.value  = Resource.Error(errorMsg)
                _dailyPrices.value    = Resource.Error(errorMsg)
                _historicalData.value = Resource.Error(errorMsg)
                _predictedData.value  = Resource.Error(errorMsg)
                _favorites.value      = Resource.Error(errorMsg)
            }

            // 天氣預警與省錢食譜獨立載入，不影響主要資料
            try {
                _weatherAlerts.value = Resource.Success(produceService.getWeatherAlerts())
            } catch (e: Exception) {
                _weatherAlerts.value = Resource.Error("天氣預警載入失敗")
            }
            try {
                _budgetRecipes.value = Resource.Success(produceService.getBudgetRecipes())
            } catch (e: Exception) {
                _budgetRecipes.value = Resource.Error("食譜推薦載入失敗")
            }
        }
    }

    /**
     * 語音搜尋：依農產品名稱查詢當日均價（供長輩語音模式使用）
     *
     * @param name 農產品名稱（例："番茄"）
     * @return 當日均價（元/公斤），查詢失敗時返回 null
     */
    suspend fun searchCropPrice(name: String): Double? {
        return try {
            val response = produceService.getDailyPrices(name, 1, 1)
            response.data.firstOrNull()?.avgPrice
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }

    /**
     * 社群零售價回報：回傳 true 表示成功，false 表示失敗
     * 成功後後端累積 5 點貢獻點數
     */
    suspend fun submitCommunityPrice(
        marketName: String,
        cropName: String,
        retailPrice: Double
    ): Boolean {
        return try {
            produceService.submitCommunityPrice(
                CommunityPriceDto(
                    cropCode = cropName,   // 後端支援名稱模糊比對
                    cropName = cropName,
                    marketName = marketName,
                    retailPrice = retailPrice
                )
            )
            true
        } catch (e: Exception) {
            e.printStackTrace()
            false
        }
    }

    /**
     * 取得使用者貢獻統計（點數、等級、回報次數）
     * @return UserStatsDto 或 null（若載入失敗）
     */
    suspend fun getUserStats(): UserStatsDto? {
        return try {
            produceService.getUserStats()
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }

    /**
     * 各市場同一農產品比價查詢
     * @param cropName 農產品名稱（例："高麗菜"）
     * @return 比價清單（依均價由低到高排序），失敗回傳空清單
     */
    suspend fun getMarketComparison(cropName: String): List<MarketCompareDto> {
        return try {
            produceService.getMarketComparison(cropName).sortedBy { it.avgPrice }
        } catch (e: Exception) {
            e.printStackTrace()
            emptyList()
        }
    }
}

File: native/iOS/UI/ProduceViewModel.swift
import Foundation
import Combine

// =============================================================================
// ProduceViewModel.swift — MVVM 架構中的 ViewModel 層
//
// 架構說明：
//   此 ViewModel 遵循 MVVM 設計模式：
//   - View (SwiftUI) 只負責顯示，不包含任何業務邏輯
//   - ViewModel 持有所有「狀態」(@Published)，並向 ProduceService 發出請求
//   - @MainActor 確保所有 @Published 變數的更新都在主執行緒 (UI Thread) 執行，
//     避免因跨執行緒更新 UI 導致 Runtime Crash
//
// 資料流向：
//   View → ViewModel.fetchDashboardData() → ProduceService (async/await) → 後端 API
//   → [解析成 DTO] → ViewModel.@Published 更新 → View 自動重新渲染
// =============================================================================

// MARK: - 資源狀態包裝器
/// Resource<T> 是三態狀態機，代表任一非同步資料請求的三種可能狀態：
/// - loading：正在載入中，UI 顯示骨架屏 (Skeleton Loader)
/// - success：載入成功，UI 顯示實際資料
/// - failure：載入失敗，UI 顯示錯誤訊息 + 重試按鈕
enum Resource<T> {
    case loading
    case success(T)
    case failure(Error)
}

// MARK: - 圖表專用本地 DTO
/// 歷史價格資料點，供 Swift Charts 折線圖使用
/// 注意：id 使用 UUID() 是因為 Charts 需要 Identifiable，
///      後端回傳的歷史資料沒有唯一 id 欄位
struct HistoricalPriceDto: Identifiable, Decodable {
    let id = UUID()
    let date: String     // 日期字串 (例："2024-01-15")
    let avgPrice: Double // 該日均價 (元/公斤)

    // CodingKeys 確保 UUID 的 id 不被 JSONDecoder 嘗試解析 (因為 JSON 中沒有 id 欄位)
    enum CodingKeys: String, CodingKey {
        case date, avgPrice
    }
}

/// 價格預測資料點，供折線圖顯示「未來預測」段落使用
/// 此 struct 為純本地計算，不從後端直接解析
struct PricePredictionDto: Identifiable {
    let id = UUID()
    let date: String             // 預測日期標籤 (例："Next Day")
    let predictedPrice: Double   // 預測均價 (元/公斤)
}

// 注意：FavoriteAlertDto 已移至 ProduceDto.swift 集中管理，
//       此處不再重複定義，以避免 Swift 編譯器 "redefinition" 錯誤。

// MARK: - ViewModel
/// ProduceViewModel 是所有 UI 畫面的核心資料中心
/// 使用 @MainActor 確保執行緒安全
@MainActor
class ProduceViewModel: ObservableObject {

    // MARK: Published States
    // 每個 @Published 對應一個 API 請求的三態狀態
    // 當任一狀態改變，SwiftUI 會自動重新渲染有訂閱此狀態的 View

    /// 我的收藏清單 + 是否達到目標提醒價格
    @Published var favorites: Resource<[FavoriteAlertDto]> = .loading

    /// 今日農產品批發價格列表 (分頁，預設第 1 頁，每頁 20 筆)
    @Published var dailyPrices: Resource<[ProduceDto]> = .loading

    /// 今日交易量前 10 名的熱門農產品
    @Published var topVolume: Resource<[ProduceDto]> = .loading

    /// 單日漲幅超過 50% 的價格異常警告清單
    @Published var anomalies: Resource<[PriceAnomalyDto]> = .loading

    /// 高麗菜 (LA1) 歷史 30 天價格資料，供折線圖展示
    @Published var historicalData: Resource<[HistoricalPriceDto]> = .loading

    /// 明日預測價格，由 7 日均線趨勢計算得出
    @Published var predictedData: Resource<[PricePredictionDto]> = .loading

    // MARK: Dependencies
    /// 透過 Protocol Injection 注入服務，方便單元測試時替換成 MockProduceService
    private let produceService: ProduceServiceProtocol

    // MARK: Init
    /// 初始化時自動觸發首頁資料載入
    /// - Parameter service: 實作 ProduceServiceProtocol 的服務實例，預設為正式 API
    init(service: ProduceServiceProtocol = ProduceService.shared) {
        self.produceService = service
        Task {
            // 在初始化完成後立即在背景非同步載入資料
            await fetchDashboardData()
        }
    }

    // MARK: - 載入首頁所有資料
    /// 並行或循序載入首頁所需的所有資料區塊
    /// 每個 do-catch 區塊獨立處理錯誤，確保某一 API 失敗不影響其他區塊
    func fetchDashboardData() async {
        // 重置所有狀態為 loading，讓 UI 顯示骨架屏
        self.anomalies = .loading
        self.topVolume = .loading
        self.dailyPrices = .loading
        self.historicalData = .loading
        self.predictedData = .loading
        self.favorites = .loading

        // 載入「價格異常警告」：漲幅超過 50% 的農產品
        do {
            let anomaliesData = try await produceService.getPriceAnomalies()
            self.anomalies = .success(anomaliesData)
        } catch {
            self.anomalies = .failure(error)
        }

        // 載入「今日熱門交易」：交易量前 10 名
        do {
            let topVolumeData = try await produceService.getTopVolumeCrops()
            self.topVolume = .success(topVolumeData)
        } catch {
            self.topVolume = .failure(error)
        }

        // 載入「今日菜價」：第 1 頁，每頁 20 筆
        do {
            let response = try await produceService.getDailyPrices(keyword: "", page: 1, pageSize: 20)
            self.dailyPrices = .success(response.data)
        } catch {
            self.dailyPrices = .failure(error)
        }

        // 載入「高麗菜 7 日趨勢圖表」
        // 選擇高麗菜 (LA1) 作為首頁示範，因為是台灣消費量最大的葉菜類
        do {
            let history = try await produceService.fetchPriceHistory(produceId: "LA1")
            self.historicalData = .success(history)

            // 根據趨勢方向計算明日預測價格：
            //   Up → 明日預測比近 7 日均價高 5%
            //   Down / Stable → 明日預測比近 7 日均價低 5%
            let forecast = try await produceService.getForecast(produceId: "LA1")
            let nextDayPrice = forecast.trend == "Up"
                ? forecast.recentAverage * 1.05
                : forecast.recentAverage * 0.95
            self.predictedData = .success([
                PricePredictionDto(date: "Next Day", predictedPrice: nextDayPrice)
            ])
        } catch {
            self.historicalData = .failure(error)
            self.predictedData = .failure(error)
        }

        // 載入「我的收藏」：使用者追蹤的農產品及是否達到目標價
        do {
            let favoritesData = try await produceService.getFavorites()
            self.favorites = .success(favoritesData)
        } catch {
            self.favorites = .failure(error)
        }
    }
}
